### ISSUE TITLE
model parallelsim on TPU v2

### ISSUE THREAD

**Issue by goexle** [sanchit-gandhi/whisper-jax#73]
Thanks for creating this repository!

I am trying to get the whipser-jax large-v2 model running on a TPU v2.
But I am running into out-of-memory issues.
Did you observe that it's necessary to run whisper-jax large-v2 with model parallelism on a TPU v2?

**Comment by iamianM** [comment]
Yeah, I'm having the same issue. Would be amazing to be able to run the large v2 model on a v2 tpu. Does anyone know if it's possible?

**Comment by sanchit-gandhi** [comment]
Hey @goexle and @iamianM! Super cool to hear that you're wanting to run Whisper JAX on TPU v2! Are you hitting OOM's when instantiating the model? Or when running it with `batch_size=8`? (the minimum per-device batch size is 1, so the minimum global batch size is 8 * 1 on a TPU v2)

You can certainly try using the model with model parallelism through pjit - I did some benchmarks using a v4-8 and it's way faster than PyTorch on an A100, but slower than doing pure data parallelism through pmap:

<img width="1313" alt="Screenshot 2023-05-05 at 15 29 05" src="https://user-images.githubusercontent.com/93869735/236486654-8fa99dd2-cbd0-4495-b8d6-80bf00f855a3.png">

I would recommend using the PALM logical axis rules:
```python
# 2D parameter and activation partitioning from PALM
logical_axis_rules_palm = [
    ("batch", None),
    ("mlp", "data"),
    ("heads", "data"),
    ("vocab", None),
    ("embed", "model"),
    ("embed", "model"),
    ("joined_kv", None),
    ("kv", None),
    ("length", None),
    ("num_mel", None),
    ("channels", None)
]
```
These worked best for me. Then you can try running the model with `num_mp_partitions=1` and `num_mp_partitions=2` ... to find the optimal number of model partitions for your hardware: https://github.com/sanchit-gandhi/whisper-jax#pipeline

**Comment by sanchit-gandhi** [comment]
Did you have any luck here @goexle @iamianM?

**Comment by shivashankar6146** [comment]
> Hey @goexle and @iamianM! Super cool to hear that you're wanting to run Whisper JAX on TPU v2! Are you hitting OOM's when instantiating the model? Or when running it with `batch_size=8`? (the minimum per-device batch size is 1, so the minimum global batch size is 8 * 1 on a TPU v2)
> 
> You can certainly try using the model with model parallelism through pjit - I did some benchmarks using a v4-8 and it's way faster than PyTorch on an A100, but slower than doing pure data parallelism through pmap:
> 
> <img alt="Screenshot 2023-05-05 at 15 29 05" width="1313" src="https://user-images.githubusercontent.com/93869735/236486654-8fa99dd2-cbd0-4495-b8d6-80bf00f855a3.png">
> 
> I would recommend using the PALM logical axis rules:
> 
> ```python
> # 2D parameter and activation partitioning from PALM
> logical_axis_rules_palm = [
>     ("batch", None),
>     ("mlp", "data"),
>     ("heads", "data"),
>     ("vocab", None),
>     ("embed", "model"),
>     ("embed", "model"),
>     ("joined_kv", None),
>     ("kv", None),
>     ("length", None),
>     ("num_mel", None),
>     ("channels", None)
> ]
> ```
> 
> These worked best for me. Then you can try running the model with `num_mp_partitions`∈ `{1, 2, 4, ...}` to find the optimal number of model partitions for your hardware: https://github.com/sanchit-gandhi/whisper-jax#pipeline

@sanchit-gandhi Thanks for sharing this :+1: `batch_size=8` and `num_mp_partitions=8` works great for me on a v2-8
